include/unifex/let_value_with_stop_token.hpp

```diff
@@ -0,0 +1,178 @@
+        (requires receiver<Receiver, Error>)
+    void set_error(Error&& error) noexcept {
+        unifex::set_error(std::move(receiver_), (Error&&)error);
+    }
+
+    void set_done() noexcept {
+        unifex::set_done(std::move(receiver_));
+    }
+
+    friend inplace_stop_token tag_invoke(tag_t<get_stop_token>, const type& r) noexcept {
+        return r.op_->stop_token_;
+    }
+
+    template(typename CPO, typename Self)
+        (requires
+            is_receiver_query_cpo_v<CPO> AND
+            same_as<Self, type>)
+    friend auto tag_invoke(CPO cpo, const Self& self)
+        noexcept(is_nothrow_callable_v<CPO, const Receiver&>)
+        -> callable_result_t<CPO, const Receiver&> {
+        return cpo(self.receiver_);
+    }
+
+private:
+    Operation* op_;
+    UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+};
+
+template<typename SuccessorFactory>
+struct _stop_token_sender {
+    class type;
+};
+
+template<typename SuccessorFactory>
+using stop_token_sender = typename _stop_token_sender<SuccessorFactory>::type;
+
+template<typename SuccessorFactory>
+class _stop_token_sender<SuccessorFactory>::type {
+    template<typename... Values>
+    using additional_arguments = type_list<type_list<Values...>>;
+
+public:
+    using InnerOp = std::invoke_result_t<SuccessorFactory, inplace_stop_token&>;
+
+    template<template<typename...> class Variant, template<typename...> class Tuple>
+    using value_types = sender_value_types_t<InnerOp, Variant, Tuple>;
+
+    template<template<typename...> class Variant>
+    using error_types = sender_error_types_t<InnerOp, Variant>;
+
+    static constexpr bool sends_done = sender_traits<InnerOp>::sends_done;
+
+    template<typename SuccessorFactory2>
+    explicit type(SuccessorFactory2&& func) : func_((SuccessorFactory2&&)func)
+    {}
+
+    template(typename Self, typename Receiver)
+        (requires same_as<remove_cvref_t<Self>, type> AND receiver<Receiver>)
+    friend auto tag_invoke(tag_t<unifex::connect>, Self&& self, Receiver&& r)
+        noexcept(
+            std::is_nothrow_invocable_v<member_t<Self, SuccessorFactory>, inplace_stop_token&> &&
+            std::is_nothrow_constructible_v<remove_cvref_t<Receiver>, Receiver>)
+        -> operation<member_t<Self, SuccessorFactory>, Receiver> {
+        return operation<member_t<Self, SuccessorFactory>, Receiver>(
+            static_cast<Self&&>(self).func_, static_cast<Receiver&&>(r));
+    }
+
+private:
+    UNIFEX_NO_UNIQUE_ADDRESS SuccessorFactory func_;
+};
+
+template<typename SuccessorFactory, typename Receiver>
+struct _stop_token_operation<SuccessorFactory, Receiver>::type {
+
+    template <typename SuccessorFactory2, typename Receiver2>
+    type(SuccessorFactory2&& func, Receiver2&& r) :
+        func_{(SuccessorFactory2&&)func},
+        stop_token_{stop_token_adapter_.subscribe(get_stop_token(r))},
+        innerOp_(
+              unifex::connect(
+                ((SuccessorFactory&&)func_)(stop_token_),
```

Comment:
You are invoking the function immediately on connect() here, and this in turn requires calling `.subscribe()` on the adapter which in turn (potentially) registers a callback on parent receiver's stop-token.

Normally, I try to avoid registering a stop-callback until start() is called, I think mainly just in case the operation is never started and so you can avoid the synchronisation needed to register the callback, but also because deferring it until start() means we often can avoid needing to store an extra bool/optional that keeps track of whether or not the stop-callback is registered (e.g. if the destructor is called without calling start()).

However, doing this would mean you couldn't get the stop-token from the adapter until start(), requiring you to defer calling the function until start(), which would in turn mean deferring calling connect() on the returned sender until start().

Other let algorithms generally seem to call the lambda passed to them at connect() time rather than at start() time so making this change would be inconsistent with them.

You may have other options if you avoid using the adapter class and instead just implement different specialisations of the operation-state based on the type of the stop-token from the parent receiver.

Reply:
I'm not sure what the right approach here should be. From my perspective, it makes sense to invoke the lambda in the call to connect() because that's required in order to connect the operation chain. In the case that lambda invocation is delayed to start() that introduces strange behavior with upstream senders where calls to connect() with them will be delayed until start() is called on this operation state. 

I attempted to remove the inplace_stop_token_adapter, but I ran into the problem where the Sender's value types are stored on the type of the Sender which requires knowing which Receiver will eventually be connected at the time of the Sender's construction. I think we said that one way around this was storing the value types on the operation state, but that seems like a large change. I'm open to exploring other ways to remove the stop_token_adapter from this algorithm though.

Reply:
One approach is to provide different specialisations of the operation-state based on what the receiver's stop-token type is.
e.g.
```c++
// generic one that adapts
template<typename Func, typename Receiver, typename StopTokenType, typename = void>
struct operation_state {
  struct receiver {
    operation_state& op_;
    
    template<typename... Values>
    void set_value(Values&&... values) noexcept {
      op_.stop_callback_.destroy();
       try { unifex::set_value(std::move(op_.receiver_), (Values&&)values...); }
       catch (...) { unifex::set_error(std::move(op_.receiver_), std::current_exception()); }
    }
   
    // etc. for set_error/set_done and also for other queries
    
    friend inplace_stop_token tag_invoke(tag_t<get_stop_token>, const receiver& r) {
      return op.stop_source_.get_token();
    }
  };
  
  void start() noexcept {
    stop_callback_.construct(get_stop_token(receiver_), detail::forward_stop_request_to_inplace_stop_source{stop_source_});
    unifex::start(sourceOp_);
  }
  
  using callback_type = typename StopTokenType::template callback_type<detail::forward_stop_request_to_inplace_stop_source>;
  
  using Source = invoke_result_t<Func, inplace_stop_token>;
  
  Receiver receiver_;
  Func func_;
  inplace_stop_source stop_source_;
  manual_lifetime<callback_type> stop_callback_;
  connect_result_t<Source, receiver> sourceOp_;
};

template<typename Func, typename Receiver, typename StopTokenType>
struct operation_state<Func, Receiver, StopTokenType, std::enable_if_t<is_stop_never_possible_v<StopTokenType>>> {

  template<typename Func2>
  operation_state(Func2&& func, Receiver&& r)
  : func_((Func2&&)func)
  , sourceOp_(unifex::connect(func_(inplace_stop_token()), (Receiver&&)r))
  {}

  void start() noexcept {
    unifex::start(sourceOp_);
  }

  using Source = invoke_result_t<Func, inplace_stop_token>;

  Func func_;
  connect_result_t<Source, Receiver> sourceOp_;
};

template<typename Func, typename Receiver>
struct operation_state<Func, Receiver, inplace_stop_token, void> {

  template<typename Func2, typename Receiver2>
  operation_state(Func2&& f, Receiver2&& r)
  : func_((Func2&&)f)
  , sourceOp_(make_source_op((Receiver2&&)r, get_stop_token(r)))
  {}

  void start() noexcept {
    unifex::start(sourceOp_);
  }

  using Source = invoke_result_t<Func, inplace_stop_token>;

  template<typename Receiver2>
  Source make_source_op(Receiver2&& r, inplace_stop_token st) {
    return unifex::connect(((Func&&)func_)(std::move(st)), (Receiver2&&)r);
  }

  Func func_;
  connect_result_t<Source, Receiver>
};
```